File: MiniProject/MoviewBoja/app/src/main/java/com/example/lkoon/movieboja/view/MainActivity.java
package com.example.lkoon.movieboja.view;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;

import com.example.lkoon.movieboja.R;

public class MainActivity extends AppCompatActivity {
    Button btnBulletin;
    Button btnInfo;
    Button btnChat;
    Button btnLocal;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        btnChat =(Button)findViewById(R.id.btnChat);
        btnLocal =(Button)findViewById(R.id.btnLocal);
        btnInfo = (Button)findViewById(R.id.btnInfo);
        btnChat.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(getApplicationContext(),ChatActivity.class);
                startActivity(intent);
            }
        });
        btnLocal.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                Intent intent = new Intent(getApplicationContext(),LocalActivity.class);
                startActivity(intent);
            }
        });
        btnInfo.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(getApplicationContext(),MovieInfoActivity.class);
                startActivity(intent);
            }
        });

        initPermission();
    }

    private void initPermission() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)!= PackageManager.PERMISSION_GRANTED) {

            // 이 권한을 필요한 이유를 설명해야하는가?
            if (ActivityCompat.shouldShowRequestPermissionRationale(this,Manifest.permission.ACCESS_FINE_LOCATION)) {

                // 다이어로그같은것을 띄워서 사용자에게 해당 권한이 필요한 이유에 대해 설명합니다

                // 해당 설명이 끝난뒤 requestPermissions()함수를 호출하여 권한허가를 요청해야 합니다

            } else {

                ActivityCompat.requestPermissions(this, new String[] {Manifest.permission.ACCESS_FINE_LOCATION
                }, 100);

                // 필요한 권한과 요청 코드를 넣어서 권한허가요청에 대한 결과를 받아야 합니다

            }
        }
    }
}

File: TreadPractice/BasicThreadAndAsync/app/src/main/java/com/lbc/practice/threadpractice/MainActivity.kt
package com.lbc.practice.threadpractice

import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.view.View
import android.widget.TextView
import kotlinx.android.synthetic.main.activity_main.*
import com.lbc.practice.threadpractice.R.id.textView


class MainActivity : AppCompatActivity() {

    var loop = true
    var isFirst = true
    var isRun = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        var thread = MyThread()

        btn_start.setOnClickListener(View.OnClickListener {
            if (isFirst) {
                isFirst = false;
                isRun = true;
                thread.start();
            } else {
                isRun = true;
            }
        })

        btn_stop.setOnClickListener(View.OnClickListener {
            isRun = false;
        })

        btn_move.setOnClickListener(View.OnClickListener {
            val intent = Intent(this, AsyncPracticeActivity::class.java)
            startActivity(intent)
        })
    }

    internal inner class MyThread : Thread() { //MainThread를 이용하여 UI에 접근

        var handler: Handler = object : Handler() {
            override fun handleMessage(msg: Message) {
                if (msg.what == 1) {  //메시지의 what 값에 따라 텍스트 변경
                    textView.text = msg.arg1.toString()
                } else if (msg.what == 2) {
                    textView.text = msg.obj as String
                }
            }
        }
        override fun run() {
            try {
                var count = 10
                while (loop) {
                    Thread.sleep(1000)
                    if (isRun) {
                        count--
                        var message = Message()
                        message.what = 1   //what값
                        message.arg1 = count
                        handler.sendMessage(message)
                        if (count == 0) {
                            message = Message()
                            message.what = 2  //what값
                            message.obj = "Finish"
                            handler.sendMessage(message)
                            loop = false
                        }
                    }
                }
            } catch (e: Exception) {

            }

        }
    }
}
